app/controllers/api/v1/journeys_controller.rb

```diff
@@ -0,0 +1,124 @@
+      end
+
+      def journeys
+        @journeys ||= move.journeys.accessible_by(current_ability).default_order
+      end
+
+      def journey
+        @journey ||= if action_name == 'create'
+                       new_journey
+                     else
+                       find_journey
+                     end
+      end
+
+      def new_journey
+        Journey.new(new_journey_attributes)
+      end
+
+      def find_journey
+        move.journeys.find(params.require(:id)).tap do |journey|
+          raise CanCan::AccessDenied.new('Not authorized', :manage, Journey) unless current_ability.can?(:manage, journey)
+        end
+      end
+
+      def validate_params
+        Journeys::ParamsValidator.new(data_params).validate!(action_name.to_sym)
+      end
+
+      def data_params
+        @data_params ||= params.require(:data)
+      end
+
+      def new_journey_params
+        @new_journey_params ||= data_params.permit(PERMITTED_NEW_JOURNEY_PARAMS).to_h
+      end
+
+      def new_journey_attributes
+        # NB: we are calling dup() to avoid mutating the underlying params object
+        @new_journey_attributes ||= new_journey_params[:attributes].dup.tap do |attribs|
+          attribs.merge!(
+            move: move,
+            supplier: current_user.owner, # NB: using the logged in account as the supplier
+            client_timestamp: Time.zone.parse(attribs.delete(:timestamp)),
+            from_location: find_location(new_journey_params.dig(:relationships, :from_location, :data, :id)),
+            to_location: find_location(new_journey_params.dig(:relationships, :to_location, :data, :id)),
+          )
+        end
+      end
+
+      def find_location(location_id)
+        # Finds the referenced location or throws an ActiveModel::ValidationError (which will render as 422 Unprocessable Entity)
+        location = Location.find_or_initialize_by(id: location_id)
+        unless location.persisted?
+          location.errors.add(:location, "reference was not found id=#{location_id}")
+          raise ActiveModel::ValidationError.new(location)
+        end
+        location
+      end
+
+      def update_journey_params
+        @update_journey_params ||= data_params.permit(PERMITTED_UPDATE_JOURNEY_PARAMS).to_h
+      end
+
+      def update_journey_attributes
+        # NB: we are calling dup() to avoid mutating the underlying params object
+        @update_journey_attributes ||= update_journey_params[:attributes].dup.tap do |attribs|
+          attribs.delete(:timestamp) # throw the timestamp away for updates
+        end
+      end
+
+      def create_event
+        # Logs the event for posterity and the immutable event log
+        Event.create!(
+          event_name: action_name,
+          eventable: journey,
+          client_timestamp: Time.zone.parse(params.dig(:data, :attributes, :timestamp)),
+          details: { data: data_params, supplier_id: current_user.owner.id },
```

Comment:
Does `owner` always get set for the current user? My understanding was that the owner would be a supplier linked to a specific oauth token, but if the front end also makes use of this API then that wouldn't have an associated supplier owner, and this could break 🤔 Maybe that's something we could guard against here?

Reply:
oops, you are quite right - good spot 👍 The front end shouldn't be creating any journeys; however, I'll update it so the FE can specify an optional supplier_id if required.

Reply:
now fixed 👍